src/node_modules/_file-explorer.js

```diff
@@ -0,0 +1,292 @@
+var setval = require('setval')
+var yo = require('yo-yo')
+var csjs = require('csjs-inject')
+var Treeview = require('ethereum-remix').ui.TreeView
+
+var css = csjs`
+  .treeview {
+    display: block;
+    margin-top: 7px;
+    width: 20%;
+  }
+  .folder,
+  .file {
+    font-size: 14px;
+  }
+  .hasFocus {
+    background-color: #F4F6FF;
+  }
+  .rename {
+    background-color: white;
+  }
+  .remove,
+  .save {
+    align-self: center;
+    margin-right: 3%;
+  }
+  .activeMode {
+    display: flex;
+    justify-content: space-between;
+  }
+`
+module.exports = fileExplorer
+
+function fileExplorer (container, files, editor) {
```

Comment:
This looks like what we are currently doing:
https://github.com/ethereum/browser-solidity/blob/master/src/app.js#L573
-> `debugAPI` contains only what is needed by the debugger.
the same here for `rendererAPI`
https://github.com/ethereum/browser-solidity/blob/master/src/app.js#L642

Except that we currently have to pass also events. Using your stuff would allow to mix API stuff and event as even the API is managed by callback in your proposal.

Basically, 
What we do here in app.js
https://github.com/ethereum/browser-solidity/blob/master/src/app.js#L642 
to
https://github.com/ethereum/browser-solidity/blob/master/src/app.js#L607

would be replaced by
```

var api = fileExplorer() // or `new FileExplorer()`
api.on('list', function (args, response) { response(null, files.list()) })
api.on('remove', function (args, response) { response(null, files.remove(args.path)) })

files.event.register('fileRenamed', function (oldName, newName) {
  api.emit('fileRenamed', { oldName, newName }, function response (err, data) { })
})

// ...
```

So the idea is interesting,
Have to think about it more ....
I think in a way I like having to pass either event and/or specific API needed to the ctor.

In your case you would have this huge block in app.js
```
var api = fileExplorer() // or `new FileExplorer()`
api.on('list', function (args, response) { response(null, files.list()) })
api.on('remove', function (args, response) { response(null, files.remove(args.path)) })

files.event.register('fileRenamed', function (oldName, newName) {
  api.emit('fileRenamed', { oldName, newName }, function response (err, data) { })
})

```


That's is still messy in a way.

I think there's (a lot of) room for improvement in in the actual remix code, we need to think about it more.
For example your solution does not clean app.js much more.
and I think that using the constructor is a good thing:
I would prefer to have 
``fileExplorer(apiDef, event1, event2)`` than 
``var api = fileExplorer()``
and the mixed api/event implementation below.

so roughly, your solution has the same idea but enforce the use of callback through this 'proxy' API.

I would keep like it is now in the debugger/renderer ...
``fileExplorer(apiDef, event1, event2)``
 - Registering to event1 and event2 in the ctor.
 - As for me, you are free to do as you feel for the apiDef (if you want to manage it by emitting/registering that's fine). That could be a first approach.

and also we should I think do that (in another PR): moving debugAPI, rendererAPI into another file, that'll free up app.js.